pandas/core/arrays/categorical.py

```diff
@@ -1712,7 +1711,7 @@ def __len__(self):
 
     def __iter__(self):
         """Returns an Iterator over the values of this Categorical."""
-        return iter(self.get_values())
+        return iter(self.get_values().tolist())
```

Comment:
Is there a reason this does `self.get_values().tolist()` instead of directly `self.tolist()` ?

Reply:
I think to avoid an infinite loop.

`Categorical.tolist` calls `tolist` which calls `Categorical.__iter__`.

Need to do `get_values` break that cycle / extract the python scalar types.

Reply:
It just feels that we are doing to much work. To iterate, we first create an array, iterate through it to create a list, and then iterate through the list ..

Reply:
So get_values either returns a array or a Datetime(like)Index. For the array we can just iterate over it, but for the Datetime(like)Index I think as well . The tolist implementation there is `list(self.astype(object))`

Reply:
Sorry, that previous comment was of course wrong, as that is exactly checking the `get_values().tolist()`

Reply:
I'm not sure if there's a strictly better way. If we want iter(Categorical) to not immediately do any work, then we do something like

```python
    def iter2(self):
        for i in range(len(self)):
            yield asscalar(self[i])
```

So while `cat.iter2()` is instant, actually consuming the iterator is slow. 22 seconds for a length-2,000,000 array, vs. 11 ms for the same array with the current implementation.

Am I missing another way, that doesn't rely on many calls to __getitem__?

Reply:
More correct comment: I think we can leave out the `tolist`, and do `iter(self.get_values()`

Or alternatively move the logic to `__iter__` (using `get_values`) and let `tolist` just call `list(iter)`?

Reply:
We need to call `tolist()` on the NumPy array from `.get_values()` so that the scalars are converted to Python types.

But yes, moving the logic sounds fine.

Reply:
Ah, yes, therefore we need the tolist. 
But then moving the logic is maybe not best, as then `tolist` would consume an iterator coming from a list .. 

It's all just complex with the different unboxing depending on the type :-)

Reply:
Simplified slightly if you want to take one last look. Basically, we didn't have to worry about the different unboxing for different types, since `Categorical.get_values()` with datetimes returns a DatetimeIndex, and when we `.tolist()` on that we get the right unboxing.

Reply:
Looks good!

Reply:
this is the whole ``.get_values()`` fiasco (which I created a while back :). I think should be addressed sooner rather than later.

Reply:
Any issues with the current implementation? Currently there's a tradeoff between how expensive `it = iter(Categorical)` is vs. `next(it)`.

Somewhere, we have to go from numpy scalars to Python scalars. The fastest way to do that is with `tolist()`, but that has upfront memory overhead. We could avoid that by doing it in the `next`, but that has time overhead for

1. converting to Python scalars elementwise
2. An extra call to CategoricalIndex.categories.__getitem__ per element

so that `next(it)` become *much* slower. I don't think there's much more to be done right now.